futures-util/src/stream/stream/filter.rs

```diff
@@ -103,34 +66,32 @@ impl<St, Fut, F> FusedStream for Filter<St, Fut, F>
 
 impl<St, Fut, F> Stream for Filter<St, Fut, F>
     where St: Stream,
-          F: FnMut(&St::Item) -> Fut,
+          F: for<'a> FnMut1<&'a St::Item, Output=Fut>,
           Fut: Future<Output = bool>,
 {
     type Item = St::Item;
 
+    #[project]
     fn poll_next(
-        mut self: Pin<&mut Self>,
+        self: Pin<&mut Self>,
         cx: &mut Context<'_>,
     ) -> Poll<Option<St::Item>> {
-        loop {
-            if self.pending_fut.is_none() {
-                let item = match ready!(self.as_mut().stream().poll_next(cx)) {
-                    Some(e) => e,
-                    None => return Poll::Ready(None),
-                };
-                let fut = (self.as_mut().f())(&item);
-                self.as_mut().pending_fut().set(Some(fut));
-                *self.as_mut().pending_item() = Some(item);
+        #[project]
+        let Filter { mut stream, f, mut pending_fut, pending_item } = self.project();
+        Poll::Ready(loop {
+            if let Some(fut) = pending_fut.as_mut().as_pin_mut() {
+                if ready!(fut.poll(cx)) {
+                    pending_fut.set(None);
```

Comment:
```suggestion
                let yield_item = ready!(fut.poll(cx));
                pending_fut.set(None);
                if yield_item {
```
I mean "if fut.poll returns Ready", not "if ready!(fut.poll) returns true" (I'm sorry if the comment was hard to understand)